texar/evals/bleu_tool.py

```diff
@@ -0,0 +1,217 @@
+        translation_corpus: list of translations to score. Each translation
+            should be tokenized into a list of tokens.
+        max_order: Maximum n-gram order to use when computing BLEU score.
+        use_bp: boolean, whether to apply brevity penalty.
+
+    Returns:
+        BLEU score.
+    """
+    reference_length = 0
+    translation_length = 0
+    bp = 1.0
+    geo_mean = 0.0
+
+    matches_by_order = [0] * max_order
+    possible_matches_by_order = [0] * max_order
+
+    for (references, translations) in zip(reference_corpus, translation_corpus):
+        reference_length += len(references)
+        translation_length += len(translations)
+        ref_ngram_counts = _get_ngrams(references, max_order)
+        translation_ngram_counts = _get_ngrams(translations, max_order)
+
+        overlap = dict(
+            (ngram, min(count, translation_ngram_counts[ngram]))
+            for ngram, count in ref_ngram_counts.items()
+        )
+
+        for ngram in overlap:
+            matches_by_order[len(ngram) - 1] += overlap[ngram]
+        for ngram in translation_ngram_counts:
+            possible_matches_by_order[len(ngram) - 1] += \
+                translation_ngram_counts[ngram]
+
+    precisions = [0.0] * max_order
+    smooth = 1.0
+
+    for i in range(max_order):
+        if possible_matches_by_order[i] > 0:
+            precisions[i] = matches_by_order[i] / possible_matches_by_order[i]
+            if matches_by_order[i] > 0:
+                precisions[i] = (matches_by_order[i] /
+                                 possible_matches_by_order[i])
+            else:
+                smooth *= 2
+                precisions[i] = 1.0 / (smooth * possible_matches_by_order[i])
+        else:
+            precisions[i] = 0.0
+
+    if max(precisions) > 0:
+        p_log_sum = sum(math.log(p) for p in precisions if p)
+        geo_mean = math.exp(p_log_sum / max_order)
+
+    if use_bp:
+        ratio = translation_length / reference_length
+        if ratio <= 0:
+            bp = 0
+        elif ratio < 1.0:
+            bp = math.exp(1 - 1.0 / ratio)
+        else:
+            bp = 1.0
+
+    bleu = geo_mean * bp
+
+    return np.float32(bleu) * 100
+
+
+class UnicodeRegex:
+    """Ad-hoc hack to recognize all punctuation and symbols."""
+
+    def __init__(self):
+        punctuation = self.property_chars("P")
+        self.nondigit_punct_re = re.compile(r"([^\d])([" + punctuation + r"])")
+        self.punct_nondigit_re = re.compile(r"([" + punctuation + r"])([^\d])")
+        self.symbol_re = re.compile("([" + self.property_chars("S") + "])")
+
+    @staticmethod
+    def property_chars(prefix):
+        return "".join(
+            chr(x)
+            for x in range(sys.maxunicode)
+            if unicodedata.category(chr(x)).startswith(prefix)
+        )
+
+
+uregex = UnicodeRegex()
+
+
+def bleu_tokenize(string: str) -> List[str]:
+    r"""Tokenize a string following the official BLEU implementation.
```

Comment:
This tokenization follows `mteval-v14.pl`. The standard BLEU implementation is [multi-bleu.perl](https://github.com/moses-smt/mosesdecoder/blob/master/scripts/generic/multi-bleu.perl). Are the tokenization parts the same?

Reply:
I think `multi-bleu.perl` requires users to tokenize the text before computing BLEU score. It is also mentioned here (https://github.com/moses-smt/mosesdecoder/blob/master/scripts/generic/multi-bleu.perl#L172). The tokenization (https://github.com/moses-smt/mosesdecoder/blob/master/scripts/generic/mteval-v14.pl#L926) in `mteval-v14.pl` is pretty standard.

Reply:
There is [another version](https://github.com/moses-smt/mosesdecoder/blob/master/scripts/generic/multi-bleu-detok.perl) of `multi-bleu.perl`, which uses the internal tokenization of `mteval-v13a.pl`.

Reply:
I see. Let's put this [statement](https://github.com/moses-smt/mosesdecoder/blob/master/scripts/generic/multi-bleu.perl#L172) (adapt if needed) in our docstring to explain the use of the `bleu_tokenize` function.